File: mobile/core/i18n/src/main/kotlin/af/velro/core/i18n/Strings.kt
package af.velro.core.i18n

import android.content.Context
import af.velro.domain.Locale
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Message keys, resolved at runtime from the same JSON the backend uses.
 *
 * Android string resources would be the ordinary choice. These are not, on
 * purpose: the server sends a `message_key` with every error, and the admin
 * panel resolves the same keys. One file set for three surfaces means a key
 * added on the server cannot be missing on the phone, and a sentence is never
 * written twice.
 *
 * No user-visible literal appears anywhere else in this codebase.
 */
class Strings private constructor(
    private val translations: Map<String, String>,
    private val fallback: Map<String, String>,
    val locale: Locale,
) {

    /**
     * Resolve a key, substituting `{named}` placeholders.
     *
     * An unknown key returns the key itself rather than an empty string: a
     * screen showing `booking.label.seat` is an obvious bug report, whereas a
     * blank space is a mystery.
     */
    operator fun get(key: String, params: Map<String, Any?> = emptyMap()): String {
        val template = translations[key] ?: fallback[key] ?: return key
        if (params.isEmpty()) return template
        val resolved = withMoney(params)
        return PLACEHOLDER.replace(template) { match ->
            val name = match.groupValues[1]
            resolved[name]?.let { format(it) } ?: match.value
        }
    }

    /**
     * Minor units in, money out.
     *
     * The server's error contexts and notification payloads carry amounts the
     * only way a machine should: `<name>_minor` plus a `currency`, locale-free
     * and exact. Three sentences spliced those raw integers straight into
     * prose -- "You need at least 5000 to request a payout" for a fifty-afghani
     * minimum, and no currency anywhere -- because substitution has no idea
     * what a number means.
     *
     * So the pairs are rendered here, in the locale actually being read, and
     * both `{amount}` and `{<name>}` become real money. A sentence that
     * genuinely wants the integer can still ask for `{<name>_minor}`.
     */
    private fun withMoney(params: Map<String, Any?>): Map<String, Any?> {
        val currency = params["currency"] as? String ?: return params
        val minors = params.keys.filter { it.endsWith(MINOR_SUFFIX) }
        if (minors.isEmpty()) return params
        val out = params.toMutableMap()
        for (key in minors) {
            val minor = (params[key] as? Number)?.toLong() ?: continue
            val money = MoneyFormatter.format(minor, currency, this)
            out[key.removeSuffix(MINOR_SUFFIX)] = money
            // The commonest shape: one amount in the sentence, called `amount`
            // whatever the context calls it.
            if (minors.size == 1) out.putIfAbsent("amount", money)
        }
        return out
    }

    operator fun get(key: String, vararg params: Pair<String, Any?>): String =
        get(key, params.toMap())

    fun has(key: String): Boolean = key in translations || key in fallback

    /**
     * The sentence for a server error code.
     *
     * The code is the contract; the wording is not. An unregistered code falls
     * back to a general message rather than showing the code to a passenger.
     */
    fun forErrorCode(code: String, context: Map<String, Any?> = emptyMap()): String {
        val key = "error." + code.lowercase()
        return if (has(key)) get(key, context) else get("error.internal_error")
    }

    private fun format(value: Any?): String = when (value) {
        null -> ""
        is Number -> Numerals.format(value, locale)
        else -> value.toString()
    }

    companion object {
        // Both braces escaped. An unescaped closing brace is accepted by the
        // JVM's regex engine and rejected by Android's ICU one, so the
        // difference only shows up on a device -- never in a JVM unit test.
        private const val MINOR_SUFFIX = "_minor"
        private val PLACEHOLDER = Regex("""\{(\w+)\}""")
        private val json = Json { ignoreUnknownKeys = true }

        /** English is the fallback because it is the only file guaranteed complete. */
        private const val FALLBACK_TAG = "en"

        /**
         * A Strings built from a literal map, for tests.
         *
         * The real one reads JSON out of the APK's assets, which needs a
         * Context and therefore an instrumented test. The substitution rules
         * -- money rendering, numerals, missing keys -- are pure, and should be
         * provable without a device.
         */
        @JvmStatic
        fun of(locale: Locale, translations: Map<String, String>): Strings =
            Strings(translations, translations, locale)

        fun load(context: Context, locale: Locale): Strings {
            val fallback = read(context, FALLBACK_TAG)
            val translations =
                if (locale.tag == FALLBACK_TAG) fallback else read(context, locale.tag)
            return Strings(translations, fallback, locale)
        }

        private fun read(context: Context, tag: String): Map<String, String> =
            runCatching {
                context.assets.open("locales/$tag.json").bufferedReader().use { reader ->
                    json.parseToJsonElement(reader.readText()).jsonObject
                        .mapValues { it.value.jsonPrimitive.content }
                }
            }.getOrDefault(emptyMap())
    }
}

File: ios/VelroCore/Sources/VelroCore/Strings.swift
import Foundation

/// Message keys, resolved from the same JSON files the backend and the Android
/// app use.
///
/// The files live in `backend/resources/locales` and are bundled into the app
/// as they are, not copied: a sentence corrected there is corrected here, and
/// a key the server sends cannot be missing on the phone. No user-visible
/// literal appears anywhere else in the app.
public struct Strings: Sendable {
    public let locale: AppLocale
    private let translations: [String: String]
    private let fallback: [String: String]

    public init(locale: AppLocale, translations: [String: String], fallback: [String: String]) {
        self.locale = locale
        self.translations = translations
        self.fallback = fallback
    }

    /// Reads `<tag>.json` from the bundle, with English underneath it: English
    /// is the only file guaranteed to be complete.
    public static func load(_ locale: AppLocale, bundle: Bundle = .main) -> Strings {
        let english = read(AppLocale.english.tag, bundle: bundle)
        let own = locale == .english ? english : read(locale.tag, bundle: bundle)
        return Strings(locale: locale, translations: own, fallback: english)
    }

    public static func read(_ tag: String, bundle: Bundle) -> [String: String] {
        guard
            let url = bundle.url(forResource: tag, withExtension: "json"),
            let data = try? Data(contentsOf: url),
            let map = try? JSONDecoder().decode([String: String].self, from: data)
        else { return [:] }
        return map
    }

    /// The sentence for a key. An unknown key comes back as itself: a screen
    /// showing `auth.field.phone` is an obvious bug report, a blank is not.
    public subscript(_ key: String) -> String {
        translations[key] ?? fallback[key] ?? key
    }

    /// The sentence with `{name}` placeholders filled in.
    ///
    /// Numbers are written in the reader's digits. A `<name>_minor` amount
    /// beside a `currency` becomes real money under `{name}` -- and under
    /// `{amount}` when it is the only one -- because the server sends money
    /// the way a machine should and a sentence needs it the way a person does.
    public subscript(_ key: String, _ params: [String: Any]) -> String {
        let template = self[key]
        guard !params.isEmpty else { return template }
        let resolved = withMoney(params)
        return template.replacing(/\{(\w+)\}/) { match in
            let name = String(match.output.1)
            guard let value = resolved[name] else { return String(match.output.0) }
            return format(value)
        }
    }

    public func has(_ key: String) -> Bool {
        translations[key] != nil || fallback[key] != nil
    }

    /// The sentence for a server error code. The code is the contract; an
    /// unregistered one reads as a general failure rather than as the code.
    public func forErrorCode(_ code: String, context: [String: Any] = [:]) -> String {
        let key = "error." + code.lowercased()
        return has(key) ? self[key, context] : self["error.internal_error"]
    }

    private func withMoney(_ params: [String: Any]) -> [String: Any] {
        guard let currency = params["currency"] as? String else { return params }
        let minors = params.keys.filter { $0.hasSuffix("_minor") }
        guard !minors.isEmpty else { return params }
        var out = params
        for key in minors {
            guard let minor = Self.integer(params[key]) else { continue }
            let money = MoneyFormatter.format(minor: minor, currency: currency, strings: self)
            out[String(key.dropLast("_minor".count))] = money
            if minors.count == 1, out["amount"] == nil { out["amount"] = money }
        }
        return out
    }

    private func format(_ value: Any) -> String {
        if let number = Self.integer(value) { return Numerals.format(number, locale) }
        if let number = value as? Double { return Numerals.localise(String(number), locale) }
        return String(describing: value)
    }

    private static func integer(_ value: Any?) -> Int64? {
        switch value {
        case let number as Int64: number
        case let number as Int: Int64(number)
        case let number as Int32: Int64(number)
        default: nil
        }
    }
}
